cli/standalone.rs

```diff
@@ -30,28 +35,39 @@ const MAGIC_TRAILER: &[u8; 8] = b"d3n0l4nd";
 /// This function will try to run this binary as a standalone binary
 /// produced by `deno compile`. It determines if this is a stanalone
 /// binary by checking for the magic trailer string `D3N0` at EOF-12.
-/// After the magic trailer is a u64 pointer to the start of the JS
-/// file embedded in the binary. This file is read, and run. If no
-/// magic trailer is present, this function exits with Ok(()).
+/// The magic trailer is followed by:
+/// - a u64 pointer to the JS bundle embedded in the binary
+/// - the u64 length of the bundle
+/// - a u64 pointer to JSON metadata (serialized flags) embedded in the binary
+/// - the u64 length of the metadata
+/// These are dereferenced, and the bundle is executed under the configuration
+/// specified by the flags. If no magic trailer is present, this function
+/// exits with `Ok(())`.
 pub fn try_run_standalone_binary(args: Vec<String>) -> Result<(), AnyError> {
   let current_exe_path = current_exe()?;
 
   let mut current_exe = File::open(current_exe_path)?;
-  let trailer_pos = current_exe.seek(SeekFrom::End(-16))?;
-  let mut trailer = [0; 16];
+  current_exe.seek(SeekFrom::End(-40))?;
+  let mut trailer = [0; 40];
   current_exe.read_exact(&mut trailer)?;
-  let (magic_trailer, bundle_pos_arr) = trailer.split_at(8);
+  let (magic_trailer, rest) = trailer.split_at(8);
   if magic_trailer == MAGIC_TRAILER {
-    let bundle_pos_arr: &[u8; 8] = bundle_pos_arr.try_into()?;
-    let bundle_pos = u64::from_be_bytes(*bundle_pos_arr);
+    let (bundle_pos, rest) = rest.split_at(8);
+    let (bundle_len, rest) = rest.split_at(8);
+    let (flags_pos, rest) = rest.split_at(8);
+    let flags_len = rest;
+    let bundle_pos = u64_from_bytes(bundle_pos)?;
+    let bundle_len = u64_from_bytes(bundle_len)?;
+    let flags_pos = u64_from_bytes(flags_pos)?;
+    let flags_len = u64_from_bytes(flags_len)?;
     current_exe.seek(SeekFrom::Start(bundle_pos))?;
 
-    let bundle_len = trailer_pos - bundle_pos;
-    let mut bundle = String::new();
-    current_exe.take(bundle_len).read_to_string(&mut bundle)?;
-    // TODO: check amount of bytes read
+    let bundle = read_string_slice(&mut current_exe, bundle_pos, bundle_len)?;
+    let flags = read_string_slice(&mut current_exe, flags_pos, flags_len)?;
```

Comment:
It'd be useful for debugging to add `.context()` before `?` with description what went wrong